File: sources/feature/landingFragment/src/main/java/com/egoriku/landingfragment/common/parallax/ParallaxScrollListener.java
package com.egoriku.landingfragment.common.parallax;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.OnScrollListener;
import android.view.ViewTreeObserver.OnGlobalLayoutListener;

import java.util.ArrayList;
import java.util.List;

public class ParallaxScrollListener extends OnScrollListener implements OnGlobalLayoutListener, LifecycleObserver {

    private final List<IParallaxScrollListener> scrollListeners = new ArrayList();
    private RecyclerView recyclerView;
    private int recyclerViewHeight;

    public void initWith(RecyclerView pRecyclerView, Lifecycle lifecycle) {
        recyclerView = pRecyclerView;
        recyclerView.getViewTreeObserver().addOnGlobalLayoutListener(this);
        recyclerView.addOnScrollListener(this);
        lifecycle.addObserver(this);
    }

    public void addListener(IParallaxScrollListener IParallaxScrollListener) {
        scrollListeners.add(IParallaxScrollListener);
    }

    public int getRecyclerViewHeight() {
        return recyclerViewHeight;
    }

    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        for (IParallaxScrollListener scrollListener : scrollListeners) {
            scrollListener.onScroll();
        }
    }

    @Override
    public void onGlobalLayout() {
        int height = 0;

        if (recyclerView != null) {
            height = recyclerView.getHeight();
        }

        if (height > 0) {
            recyclerViewHeight = height;
            recyclerView.getViewTreeObserver().removeOnGlobalLayoutListener(this);

            for (IParallaxScrollListener scrollListener : scrollListeners) {
                scrollListener.onAttach();
            }
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    protected void detach() {
        for (IParallaxScrollListener scrollListener : scrollListeners) {
            scrollListener.onDetach();
        }
        recyclerView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
        recyclerView.removeOnScrollListener(this);
        scrollListeners.clear();
        recyclerView = null;
    }
}

File: sources/feature/landing/src/main/java/com.egoriku.ladyhappy.landing/common/parallax/ParallaxScrollListener.kt
package com.egoriku.ladyhappy.landing.common.parallax

import android.view.ViewTreeObserver.OnGlobalLayoutListener
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.OnLifecycleEvent
import androidx.recyclerview.widget.RecyclerView

open class ParallaxScrollListener : RecyclerView.OnScrollListener(), OnGlobalLayoutListener, LifecycleObserver {

    private val scrollListeners = mutableListOf<ParallaxScrollStateListener>()
    private var recyclerView: RecyclerView? = null

    var recyclerViewHeight = 0
        private set

    fun initWith(view: RecyclerView, lifecycle: Lifecycle) {
        recyclerView = view
        recyclerView?.viewTreeObserver?.addOnGlobalLayoutListener(this)
        recyclerView?.addOnScrollListener(this)
        lifecycle.addObserver(this)
    }

    fun addListener(scrollListenerListener: ParallaxScrollStateListener) {
        scrollListeners.add(scrollListenerListener)
    }

    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
        scrollListeners.forEach { scrollListener ->
            scrollListener.onScroll()
        }
    }

    override fun onGlobalLayout() {
        var height = 0
        if (recyclerView != null) {
            height = recyclerView!!.height
        }
        if (height > 0) {
            recyclerViewHeight = height
            recyclerView?.viewTreeObserver?.removeOnGlobalLayoutListener(this)

            scrollListeners.forEach { scrollListener ->
                scrollListener.onAttach()
            }
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    protected fun detach() {
        scrollListeners.forEach { scrollListener ->
            scrollListener.onDetach()
        }

        recyclerView?.viewTreeObserver?.removeOnGlobalLayoutListener(this)
        recyclerView?.removeOnScrollListener(this)
        scrollListeners.clear()
        recyclerView = null
    }
}